onnxruntime/core/providers/cpu/math/matmul_integer.cc

```diff
@@ -52,35 +51,67 @@ Status MatMulInteger<uint8_t, uint8_t, int32_t>::Compute(OpKernelContext* ctx) c
-        "Currently only scalar zero_point is supported. TODO: add per channel zero point support.");
-    a_offset = static_cast<int32_t>(*a_zero_point->template Data<uint8_t>());
-  }
-  if (has_b_zero_point_) {
-    auto b_zero_point = ctx->Input<Tensor>(3);
-    ORT_ENFORCE(b_zero_point->Shape().NumDimensions() == 0 ||
-        (b_zero_point->Shape().NumDimensions() == 1 && b_zero_point->Shape().GetDims().size() == 1),
-        "Currently only scalar zero_point is supported. TODO: add per channel zero point support.");
-    b_offset = static_cast<int32_t>(*b_zero_point->template Data<uint8_t>());
-  }
+  ORT_RETURN_IF_NOT(y->DataType() == DataTypeImpl::GetType<std::int32_t>());
+
+  if (a->DataType() == DataTypeImpl::GetType<std::uint8_t>() &&
+      b->DataType() == DataTypeImpl::GetType<std::uint8_t>()) {
+    // validate zero points
+    int32_t a_offset = 0;
+    int32_t b_offset = 0;
+    if (has_a_zero_point_) {
+      auto a_zero_point = ctx->Input<Tensor>(2);
+      ORT_ENFORCE(a_zero_point->Shape().NumDimensions() == 0 ||
+                      (a_zero_point->Shape().NumDimensions() == 1 && a_zero_point->Shape().GetDims().size() == 1),
+                  "Currently only scalar zero_point is supported. TODO: add per channel zero point support.");
+      a_offset = static_cast<int32_t>(*a_zero_point->template Data<uint8_t>());
+    }
+    if (has_b_zero_point_) {
+      auto b_zero_point = ctx->Input<Tensor>(3);
+      ORT_ENFORCE(b_zero_point->Shape().NumDimensions() == 0 ||
+                      (b_zero_point->Shape().NumDimensions() == 1 && b_zero_point->Shape().GetDims().size() == 1),
+                  "Currently only scalar zero_point is supported. TODO: add per channel zero point support.");
+      b_offset = static_cast<int32_t>(*b_zero_point->template Data<uint8_t>());
+    }
 
-  for (size_t i = 0; i < helper.OutputOffsets().size(); i++) {
-    GemmlowpMultiply(a->template Data<uint8_t>() + helper.LeftOffsets()[i],
-                     b->template Data<uint8_t>() + helper.RightOffsets()[i],
-                     y->template MutableData<int32_t>() + helper.OutputOffsets()[i],
-                     a_offset,
-                     b_offset,
-                     static_cast<int>(helper.M()),
-                     static_cast<int>(helper.N()),
-                     static_cast<int>(helper.K()));
+    for (size_t i = 0; i < helper.OutputOffsets().size(); i++) {
+      GemmlowpMultiply(a->template Data<uint8_t>() + helper.LeftOffsets()[i],
+                       b->template Data<uint8_t>() + helper.RightOffsets()[i],
+                       y->template MutableData<int32_t>() + helper.OutputOffsets()[i],
+                       a_offset,
+                       b_offset,
+                       static_cast<int>(helper.M()),
+                       static_cast<int>(helper.N()),
+                       static_cast<int>(helper.K()));
+    }
+  } else {
+    if (has_a_zero_point_) {
```

Comment:
also check the value... in a lot of cases we add zero point as 0.... this should also be treated as a valid case

Reply:
Good point, will do.

---
In reply to: [303684437](https://github.com/microsoft/onnxruntime/pull/1391#discussion_r303684437) [](ancestors = 303684437)